specs/advanced-data-layouts/prollytree/spec.md

```diff
@@ -0,0 +1,467 @@
+9. `target = (end - start)/d`
+10. return `p < target`
+
+### HashByteArray([]byte, offset) : Boolean
+
+*TODO: Description*
+*TODO: current implementation applies the min/maxChunkSize to offset. Do we want this?*
+- For each `byte`:
+    1. Increment `offset` by 1
+    2. Feed `byte` to rolling hash *TODO Details for hashing, salting*
+    3. Set `hash` to hashSum of rolling hash
+    4. `pattern = (1<<(15 - (offset >> 10))) - 1`
+    5. If `hash&pattern == pattern` return `true`
+- Return `false`
+
+### RollingHash(ProllyNode, index) : Boolean
+
+*TODO: Rolling hash window*
+1. Set `offset` to the number of bytes of the elements of previous keys and values
+2. If `HashByteArray(key) == true`: return `true`
+3. Add number of bytes of key to `offset`
+4. If `HashByteArray() == true`: return `true`
+5. return `false`
+
+### ShouldCreateBoundry(ProllyNode, key) : Boolean
+
+If `true`, a new ProllyNode should be created for all subsequent keys.
+
+1. Get the `ChunkConfig` by resolving the link in `node.ChunkConfig`
+1. Get the `ChunkStrategy` from the `ChunkConfig`
+2. Get the `length` from the `node.keys`
+3. If `length` is less than `ChunkConfig.MinChunkSize` return `true`
+3. If `length` is equal to `ChunkConfig.MaxChunkSize` return `true`
+4. If the `ChunkStrategy` is `KeySplitter`
+	- return WeibullThreshold(key)
+5. If the `ChunkStrategy` is `RollingHash`
+	- return RollingHash(ProllyNode, index)
+6. Return an error (unsupported chunk strategy)
+
+### RebalanceTree(Cursor) : ProllyNode
+
+The `cursor` should be pointing to the node in a tree that rebalancing should start at. The root of the `cursor` will be upated to match the changes.
+
+The returned `ProllyNode` is for the new root of the tree.
+
+TODO: Account for `cursor.parent` being null
+TODO: Advance a cursor instead of iterating through `cursor.node.keys`
+TODO: Account for `cursor.node.keys` being empty (remove from parent)
+
+- Loop
+  - If `ShouldCreateBoundry(cursor.node.chunkconfig, cursor.node, CursorGetKey(cursor))` is `false`
+	- Call `AdvanceCursor(cursor)`
+	- continue to next loop cycle
+  - Create a new `ProllyNode` `node` by cloning the current `cursor.node`
+  - Remove all keys, values, links from `cursor.node` after `cursor.index`
+  - Remove all keys, values, links from `node` before and including `cursor.index`
+  - Generate a new `CID` for `cursor.node`
+  - Set the `cursor.parent.links` at `cursor.parent.index` to the new `CID`
+  - Generate a new `CID` for the new `node`
+  - Shift all items in `cursor.parent.keys` up by `1`
+  - Shift all items in `cursor.parent.links` up by `1`
+  - Increment `cursor.parent.index` by `1`
+  - Set the `cursor.parent.links` at `cursor.parent.index` to the new `CID`
+  - Set the `cursor.parent.keys` at `cursor.parent.index` to node.keys[0]
+  - Set cursor.node to `node`
+  - Set cursor.index to `0`
+- Return `RebalanceTree(cursor.parent)`
+
+### Put(ProllyNode, key, value) : ProllyNode
```

Comment:
structure method,  and I think it's better to have another struct like ProllyTree?